x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/CIDRMatch.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.expression.function.scalar.string;
+
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Expressions;
+import org.elasticsearch.xpack.ql.expression.Expressions.ParamOrdinal;
+import org.elasticsearch.xpack.ql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.expression.predicate.logical.Or;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.util.CollectionUtils;
+
+import java.util.List;
+
+import static java.util.Collections.singletonList;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isIPAndExact;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isStringAndExact;
+
+/**
+ * EQL specific cidrMatch function
+ */
+public class CIDRMatch extends ScalarFunction {
+
+    private final Expression field;
+    private final List<Expression> addresses;
+
+    public CIDRMatch(Source source, Expression field, List<Expression> addresses) {
+        super(source, CollectionUtils.combine(singletonList(field), addresses));
+        this.field = field;
+        this.addresses = addresses;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (!childrenResolved()) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isIPAndExact(field, sourceText(), Expressions.ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        for (Expression addr: addresses) {
+            // Currently we have limited enum for ordinal numbers
+            // So just using default here for error messaging
+            resolution = isStringAndExact(addr, sourceText(), ParamOrdinal.DEFAULT);
```

Comment:
The current enum for error messages doesn't work great for variadic type of function. Open to suggestions.

Reply:
I added a method here in my wildcard PR:
https://github.com/elastic/elasticsearch/blob/d482c472d05af7220ce1896e323deb30769b05f8/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/Expressions.java#L36-L44

Reply:
Also, should these fields also be `isStringAndExact`? and an `isFoldable` check?
Here's how I approached that
https://github.com/elastic/elasticsearch/blob/d482c472d05af7220ce1896e323deb30769b05f8/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/Wildcard.java#L80-L83

Reply:
Cool! Will update after your PR merges.
The ParamOrdinal falls back to ParamOrdinal.DEFAULT after 4 params anyways. Maybe make enum longer if we still want to use enum there?

Reply:
For how to handle functions that deal with a variable list of parameters, I suggest having a look at `org.elasticsearch.xpack.sql.expression.predicate.conditional.Case` or `org.elasticsearch.xpack.sql.expression.predicate.conditional.Coalesce` since these are functions that have a variable list of parameters.

Reply:
Relying on the same def builder here that was merged with `wildcard` function for now.
Updated to align with wildcard impl usage of ParamOrdinal.